src/steps/email-links-validation.ts

```diff
@@ -71,18 +72,25 @@ export class EmailLinksValidationStep extends BaseStep implements StepInterface
       const dom = parser.parseFromString(htmlBody);
 
       const htmlUrls = dom.getElementsByTagName('a')
-                      .map(f => f.getAttribute('href'))
-                      .filter(f => f.includes('http'));
-      const plainUrls = Array.from(GetUrls(plain).values());
+                      .map((f) => { return { url: f.getAttribute('href'), type: 'HTML' }; })
+                      .filter(f => f.url.includes('http'));
+
+      const plainUrls = Array.from(GetUrls(plain).values()).map((f) => { return { url: f, type: 'Plain' }; });
 
       const urls = new Set(htmlUrls.concat(plainUrls));
 
-      const brokenUrls = await this.client.evaluateUrls(
+      const response = await this.client.evaluateUrls(
         this.sanitizeUrl(Array.from(urls.values())));
 
-      if (brokenUrls.length > 0) {
-        return this.fail('Broken links found in the email. URLs include: %s', [
-          brokenUrls.join(', '),
+      if (response.length > 0) {
+        const plain = response.filter(f => f.type === 'Plain');
+        const html = response.filter(f => f.type === 'HTML');
+        return this.fail('Broken links found in the email. URLs include: %s %s %s %s %s', [
```

Comment:
I've also added a proposal to Crank CLI to introduce a way to pass structured data from a Cog to the CLI to avoid all of this string munging: https://github.com/run-crank/cli/issues/28